### ISSUE TITLE
Validate Ticket after a long time

### ISSUE THREAD

**Issue by vicb25** [IUCrimson/AspNet.Security.CAS#12]
Am I missing something ? Once the user is authenticated it doesn't look like the ticket is ever validated again to check if the ticket/cookie is still valid (user is never prompted to login again, until the session cookie is deleted from the browser or browser is closed), unless I am not seeing something.

**Comment by davidmdem** [comment]
You're not missing anything - this is how the sample app works and is more of an issue regarding how the Cookie middleware is setup.

There are two sessions login cookies at play here though and I'm not 100% clear as to whether asking about the CAS server's cookie (`CASTGC` on your CAS server's domain) or your application's cookie (`.AspNetCore.CasSample` for the sample app).  At least in the CAS implementation I have available, once you sign into CAS, you are signed in for the duration of that browser session.

Check my response to [this issue to see if it clarifies some things](https://github.com/IUCrimson/AspNet.Security.CAS/issues/7).

With that said, there are a couple of places to set a cookie expiration for your application.

```c#
.AddCookie(cookieOptions =>
{
    cookieOptions.ExpireTimeSpan = TimeSpan.FromDays(30);
    ...
}
```

This will tell your server to expire authentication tickets after the specified timespan. I'm not sure how helpful this will be though if your CAS server keeps the user signed in for the whole session because your app will see them as signed out, send them to CAS which will transparently sign them in again (see the discussion linked above).

Where this would be more beneficial is if you wanted to keep the cookie around *longer* than a single session.  A "Remember me" type feature that would prevent them from returning to CAS every session.

If you're wanting the cookie to stick around in the browser for longer than the current session, you'll also need to set the `ExpireUtc` and `IsPersistent` property in your `AuthenticationProperties`:

```c#
[AllowAnonymous]
[HttpGet("login")]
public async Task Login(string returnUrl)
{
    var props = new AuthenticationProperties
    {
        RedirectUri = returnUrl, 
        IsPersistent = true,
        ExpiresUtc = DateTimeOffset.UtcNow.AddDays(30)
    };
    await HttpContext.ChallengeAsync("CAS", props);
}
```

**Comment by vicb25** [comment]
So the code can never react to back-end changes. 

So after the user is logged in and then the user is suspended/password changed/or hacked - the user will be able to continue using the site.

https://docs.microsoft.com/en-us/aspnet/core/security/authentication/cookie?tabs=aspnetcore2x#reacting-to-back-end-changes

We cannot validate CAS ticket as its one time ( any ideas on how to check or validate on every request) or how to solve the CAS Auth Middleware to react to back-end changes

**Comment by davidmdem** [comment]
I believe adding support for the ['renew' option](https://apereo.github.io/cas/5.0.x/protocol/CAS-Protocol-V2-Specification.html#211-parameters) is what is needed.

From the docs: 

> if this parameter is set, single sign-on will be bypassed. In this case, CAS will require the client to present credentials regardless of the existence of a single sign-on session with CAS.

Your application will still need to implement some kind of custom `ValidatePrincipal` for you to invalidate the client application cookie.  If the renew option was specified as true, the user would be prompted again by CAS after you invalidate their cookie, even in the same browser session.

Would this work?

**Comment by davidmdem** [comment]
The `Renew` and `Gateway` options were implemented in the 2.0.2 release.
